pkg/kubelet/kubelet_pods.go

```diff
@@ -973,7 +973,7 @@ func (kl *Kubelet) filterOutInactivePods(pods []*v1.Pod) []*v1.Pod {
 		}
 
 		// terminal pods are considered inactive UNLESS they are actively terminating
-		if kl.isAdmittedPodTerminal(p) && !kl.podWorkers.IsPodTerminationRequested(p.UID) {
+		if kl.isAdmittedPodTerminal(p) && !kl.podWorkers.IsPodTerminating(p.UID) {
```

Comment:
Me and @wenwu449 took a closer look don't fully understand the fix. The issue is that  prior on line 971 we have `kl.podWorkers.IsPodKnownTerminated` which uses `status.IsTerminated()` underneath (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/pod_workers.go#L439-L447).

We can consider the following 4 cases:

1. podA - `status.IsTerminated()` => false ; `status.IsTerminationRequested()` => true
2. podB - `status.IsTerminated()` => false; `status.IsTerminationRequested()` => false
3. podC - `status.IsTerminated()` => true; `status.IsTerminationRequested()` => true
4. podD - `status.IsTerminated()` => true; `status.IsTerminationRequested()` => false

Before this fix, this was result after calling `filterOutInactivePods` for each of the pods:

podA - false => added to active pods
podB - true => continue (inactive)
podC - true => continue (inactive)
podD - true => continue (inactive)

After this fix, the output is the same:

podA - false => added to active pods
podB - true => continue (inactive)
podC - true => continue (inactive)
podD - true => continue (inactive)

Line 971, of checking `kl.podWorkers.IsPodKnownTerminated(p.UID)` will basically short circut, so for podC and podD, we won't reach the new code (since they both have `status.IsTerminated()`.

So only the interesting cases are podA and podB, but there is no change output after the fix for these cases. Let me know if I'm misunderstanding something here...

Reply:
Yea, Line 971, of checking `kl.podWorkers.IsPodKnownTerminated(p.UID)` will filter all terminated pod including terminal ones(Succeed or Failed). I still have some doubts that in which condition `kl.podWorkers.IsPodTerminationRequested(p.UID)` can return false for a terminal pod(Succeed or Failed), as a terminal pod would be terminated by podworker immediately and it's podstatus would be kept until it's GCed.  so` if kl.isAdmittedPodTerminal(p) && !kl.podWorkers.IsPodTerminationRequested(p.UID)`  may always false, right?

Reply:
You’re right, IsKnown will prevent the scenario I described from happening, and this code change is a no-op, and now i remember the discussion that led to the fi line.

No pod terminates immediately - in practice the worlds fastest pod deletion / termination takes about a second to propagate through the kubelet.  The source of truth for termination is the kubelet, so in normal action there should be no scenario where pod status is terminal in this loop, but it can happen if a user modifies phase, or the kubelet has a racy / incorrect update.  I’ll take a look with the reproducer in the morning.